Add JITServer support for new invokedynamic and invokehandle

dmitry-ten opened the pull request:
This enables JITServer support for #10607.
Parts of the new Method Handle code require holding
VM access in the JIT, so to make it work on JITServer,
all such code has been wrapped in front-end calls.
When executed on the server, these front-end queries will
send a message to the client, which will do the required
work while holding VM access and send the result to the server.

This commit also refactors
`TR_ResolvedJ9JITServerMethod::getResolvedHandle(Dynamic)Method` to
make it compatible with the new method handle implementation
and adds a server-side version of `createResolvedMethod(WithSignature)`
that takes `TR_ResolvedJ9JITServerMethodInfo` as parameter, allowing
to create a resolved method without a remote call if we already
prefetched the method info.

dmitry-ten commented:
@nbhuiyan could you take a look at this?

nbhuiyan commented on runtime/compiler/env/VMJ9.h:
```diff
@@ -777,27 +777,48 @@ class TR_J9VMBase : public TR_FrontEnd
+    *    Create and return a resolved method from member name index of an invoke cache array.
+    *    Encapsulates code that requires VM access to make JITServer support easier.
+    */
+   virtual TR_ResolvedMethod* targetResolvedMethodFromInvokeCacheArray(TR::Compilation *comp, TR_ResolvedMethod *owningMethod, uintptr_t *invokeCacheArray);
```
Perhaps we could use a more explicit name here, like `targetMethodFromInvokeCacheArrayMemberNameObj`?

dmitry-ten replied:
I agree that it's a better name

nbhuiyan commented on runtime/compiler/env/VMJ9.cpp:
```diff
@@ -4858,6 +4859,37 @@ TR_J9VMBase::targetMethodFromMethodHandle(uintptr_t methodHandle)
+   {
+   TR::VMAccessCriticalSection vmAccess(this);
+   uintptr_t arrayElementRef = (uintptr_t) getReferenceElement(*invokeCacheArray, arrayIndex);
+   return comp->getSymRefTab()->refineInvokeCacheElementSymRefWithKnownObjectIndex(originalSymRef, arrayElementRef);
```
If the refinement of the symref has to be done via a call to a J9VM helper to make this work in JITServer, we should just move the implementation of `J9::SymbolReferenceTable::refineInvokeCacheElementSymRefWithKnownObjectIndex` here.

dmitry-ten replied:
It's not that we can't do it from `J9::SymbolReferenceTable::refineInvokeCacheElementSymRefWithKnownObjectIndex`, we could, but I want to avoid placing remote calls outside of front-end queries as much as possible.
If you are fine with moving the implementation here, I'll do it.

nbhuiyan commented on runtime/compiler/env/VMJ9Server.cpp:
```diff
@@ -154,6 +161,38 @@ TR_J9ServerVM::createResolvedMethodWithSignature(TR_Memory * trMemory, TR_Opaque
+                                                 char *signature, int32_t signatureLength, TR_ResolvedMethod * owningMethod, const TR_ResolvedJ9JITServerMethodInfo &methodInfo)
+   {
+   TR_ResolvedJ9Method *result = NULL;
+   if (isAOT_DEPRECATED_DO_NOT_USE())
```
Should we be introducing more uses of a function meant to be deprecated? Would `TR::Compilation->compileRelocatableCode()` not work here?

dmitry-ten replied:
It would, I just copy-pasted this code from the non-jitserver implementation of this method.
I'll change it to `compileRelocatableCode()`.

nbhuiyan commented on runtime/compiler/env/VMJ9.h:
```diff
@@ -777,27 +777,48 @@ class TR_J9VMBase : public TR_FrontEnd
+    *    Wrapper for J9::SymbolReferenceTable::refineInvokeCacheElementSymRefWithKnownObjectIndex.
+    *    Encapsulates code that requires VM access to make JITServer support easier.
+    */
+   virtual TR::SymbolReference* refineInvokeCacheElementSymRefWithKnownObjectIndex(TR::Compilation *comp, TR::SymbolReference *originalSymRef, uintptr_t *invokeCacheArray, int32_t arrayIndex);
```
Since we only ever refine appendix object symref when invokedynamic/invokehandle is resolved, we could change this to just handle refining the appendix object and rename the function accordingly. We can get rid of the `arrayIndex` parameter that remains the same in every call, and use the `JSR292_invokeCacheArrayAppendixIndex` macro directly in this function.

dmitry-ten replied:
Yeah, that was something I was wondering about. Are there no potential future uses, where we could pass a different index to this function? If not, then I agree with your suggestion

nbhuiyan replied:
No, we will not have to pass a different index in the future unless the side table design changes

dmitry-ten commented on runtime/compiler/env/VMJ9.h:
```diff
@@ -777,27 +777,48 @@ class TR_J9VMBase : public TR_FrontEnd
+    *    Return a Known Object Table index of an invoke cache array appendix element.
+    *    Encapsulates code that requires VM access to make JITServer support easier.
+    */
+   virtual TR::KnownObjectTable::Index getKnotIndexOfInvokeCacheArrayAppendixIndex(TR::Compilation *comp, uintptr_t *invokeCacheArray);
```
I agree

dmitry-ten commented:
@nbhuiyan implemented the requested changes, ready for another review.

nbhuiyan commented on runtime/compiler/optimizer/InterpreterEmulator.cpp:
```diff
@@ -651,10 +648,7 @@ InterpreterEmulator::visitInvokehandle()
-         || comp()->isOutOfProcessCompilation()
-#endif
-      ) return; // do nothing if unresolved, is AOT or JITServer compilation
+      ) return; // do nothing if unresolved, is an AOT compilation
```
typo: missing "or" in comment

dmitry-ten commented on runtime/compiler/compile/J9SymbolReferenceTable.hpp:
```diff
@@ -108,14 +108,6 @@ class SymbolReferenceTable : public OMR::SymbolReferenceTableConnector
    TR::SymbolReference * findOrCreateHandleMethodSymbol(TR::ResolvedMethodSymbol * owningMethodSymbol, int32_t cpIndex, bool * unresolvedInCP = 0);
    TR::SymbolReference * findOrCreateHandleMethodSymbol(TR::ResolvedMethodSymbol * owningMethodSymbol, int32_t cpIndex, char *signature);
 #if defined(J9VM_OPT_OPENJDK_METHODHANDLE)
```
Right, forgot to remove it.
This also reminded me to surround code to handle new message types with ifdef.

keithc-ca replied:
Is the intent to remove this declaration in a future PR?

mpirvu replied:
> Is the intent to remove this declaration in a future PR?

This define is present in many places of the VM/GC/JIT. @babsingh may be better suited to answer about the plans around the  `J9VM_OPT_OPENJDK_METHODHANDLE` define.

babsingh replied:
The end goal is to adopt the OpenJDK MHs. Once OJDK MHs are functionally stable, performant and enabled across all OJ9 JDKs, then the OJ9 MHs will be removed along with usage of ` J9VM_OPT_OPENJDK_METHODHANDLE`.

keithc-ca replied:
I was referring to the declaration of `refineInvokeCacheElementSymRefWithKnownObjectIndex` below which doesn't appear to have a corresponding implementation.

dmitry-ten replied:
Right, I should have deleted that declaration, I don't know how it's still there.
I'll make a new PR to delete it.

dmitry-ten commented on runtime/compiler/control/JITClientCompilationThread.cpp:
```diff
@@ -42,6 +42,7 @@
 #include "runtime/JITServerIProfiler.hpp"
 #include "runtime/RelocationTarget.hpp"
 #include "env/TypeLayout.hpp"
+#include "env/JSR292Methods.h"
```
Ok, done.

dmitry-ten commented:
Addressed the most recent comments.

dmitry-ten commented:
@mpirvu could you take a look at this as well now?

nbhuiyan reviewed (approved):
Other than the unresolved [comment](https://github.com/eclipse/openj9/pull/11933#discussion_r579462884) in an earlier review round to rename `refineInvokeCacheElementSymRefWithKnownObjectIndex` to something reflecting what it does now, LGTM! My suggestion: `refineInvokeCacheAppendixObjWithKnotIndex`

dmitry-ten commented:
Renamed `refineInvokeCacheElementSymRefWithKnownObjectIndex ` to `refineInvokeCacheAppendixObjWithKnotIndex`
and fixed the error where I didn't account for the case when `getResolvedHandle(Dynamic)Method` returns NULL.

mpirvu reviewed (approved):
I only have a small observation related to marking virtual methods `override`.

For my own understanding I would like to know:
1) What is the InterpreterEmulator used for?
2) What is invokeCacheArray? The name suggests it's a cache. What does it cache? Is it per method/VM? Who populates/updates it?

TR_J9ByteCodeIlGenerator::genInvokeDynamic() sends up to 4 messages now. If we find a large frequency of messages, then there is room for improvement because all those 4 messages can be grouped together into 1.

dmitry-ten commented:
1. Interpreter Emulator is an inliner utility that walks the bytecodes of a method and creates call sites for each method call bytecode. It's used in the code size estimation part of the inliner.
2. I'll leave the `invokeCacheArray` explanation to @nbhuiyan since he understands how OpenJDK MH work better.

I understand that `genInvokeDynamic` may send a lot of messages but I haven't observed many dynamic methods called in any of the tests I ran, so the number of additional messages is not significant. This could be reduced to 1 message, if lines in `genInvokeDynamic` are moved to a front-end call but I didn't want to make one big query that would do a lot of stuff  for ilgen.

dmitry-ten commented on runtime/compiler/env/VMJ9Server.hpp:
```diff
@@ -61,8 +62,9 @@ class TR_J9ServerVM: public TR_J9VM
    virtual TR_ResolvedMethod * createResolvedMethod(TR_Memory * trMemory, TR_OpaqueMethodBlock * aMethod, TR_ResolvedMethod * owningMethod, TR_OpaqueClassBlock *classForNewInstance) override;
-   virtual TR_ResolvedMethod * createResolvedMethodWithSignature(TR_Memory * trMemory, TR_OpaqueMethodBlock * aMethod, TR_OpaqueClassBlock *classForNewInstance,
-                                                                 char *signature, int32_t signatureLength, TR_ResolvedMethod * owningMethod) override;
+   TR_ResolvedMethod * createResolvedMethod(TR_Memory * trMemory, TR_OpaqueMethodBlock * aMethod, TR_ResolvedMethod * owningMethod, const TR_ResolvedJ9JITServerMethodInfo &methodInfo, TR_OpaqueClassBlock *classForNewInstance = NULL);
+   virtual TR_ResolvedMethod * createResolvedMethodWithSignature(TR_Memory * trMemory, TR_OpaqueMethodBlock * aMethod, TR_OpaqueClassBlock *classForNewInstance, char *signature, int32_t signatureLength, TR_ResolvedMethod * owningMethod);
+   TR_ResolvedMethod * createResolvedMethodWithSignature(TR_Memory * trMemory, TR_OpaqueMethodBlock * aMethod, TR_OpaqueClassBlock *classForNewInstance, char *signature, int32_t signatureLength, TR_ResolvedMethod * owningMethod, const TR_ResolvedJ9JITServerMethodInfo &methodInfo);
```
One of them should be marked `override`, but others are not virtual methods, since they require `TR_ResolvedJ9JITServerMethodInfo` as argument.

mpirvu reviewed (approved):
LGTM

mpirvu commented:
jenkins test sanity plinuxjit,xlinuxjit,zlinuxjit jdk11

dmitry-ten commented:
Build failure due to missing #ifdef in `JITClientCompilationThread.cpp`. Weird, I thought I added it in the last push, added it now.

mpirvu commented:
jenkins test sanity plinuxjit,xlinuxjit,zlinuxjit jdk11

nbhuiyan commented:
@mpirvu
> What is invokeCacheArray? The name suggests it's a cache. What does it cache? Is it per method/VM? Who populates/updates it?

`invokeCacheArray` (also can be simply referred to as `invokeCache`) we see here is the address of a 2-element Java array containing Java object references that are used in resolution and dispatch of `invokedynamic`/`invokehandle` calls. The two objects are `memberName` objects (the actual adapter method to call), and `appendix` object, which we need to provide as the last argument for the adapter method call). Each invokedynamic/invokehandle will have its own `invokeCacheArray` entry that can be looked up using the `callSiteIndex` or `cpIndex` respectively, but they do not use the same table. For `invokedynamic`, the table to lookup the `invokeCache` entry is in `ramClass->callSites` and for `invokehandle`, it is in `ramClass->invokeCache`. In the JIT, to check if an `invokedynamic`/`invokehandle` call is resolved, we check if the corresponding `invokeCache` entry has been populated through method handle resolution. Once resolved, the there will be no further changes to the `invokeCache` entry for an `invokedynamic`/`invokehandle` call site.
